loadbalancer/zk/client.go

```diff
@@ -0,0 +1,157 @@
+package zk
+
+import (
+	"errors"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/eapache/channels"
+	"github.com/samuel/go-zookeeper/zk"
+
+	"github.com/go-kit/kit/log"
+)
+
+// DefaultACL is the default ACL to use for creating znodes
+var DefaultACL = zk.WorldACL(zk.PermAll)
+
+const (
+	// ConnectTimeout is the default timeout to establish a connection to a
+	// ZooKeeper node.
+	ConnectTimeout = 2 * time.Second
+	// SessionTimeout is the default timeout to keep the current ZooKeeper
+	// session alive during a temporary disconnect.
+	SessionTimeout = 5 * time.Second
+)
+
+// Client is a wrapper around a lower level ZooKeeper client implementation.
+type Client interface {
+	// GetEntries should query the provided path in ZooKeeper, place a watch on
+	// it and retrieve data from its current child nodes.
+	GetEntries(path string) ([]string, channels.SimpleOutChannel, error)
+	// CreateParentNodes should try to create the path in case it does not exist
+	// yet on ZooKeeper.
+	CreateParentNodes(path string) error
+}
+
+type clientConfig struct {
+	acl             []zk.ACL
+	connectTimeout  time.Duration
+	sessionTimeout  time.Duration
+	rootNodePayload [][]byte
+}
+
+type configOption func(c *clientConfig) error
+
+type client struct {
+	*zk.Conn
+	clientConfig
+	Eventc <-chan zk.Event
+	quit   chan struct{}
+}
+
+// WithACL returns a configOption specifying a non-default ACL.
+func WithACL(acl []zk.ACL) configOption {
+	return func(c *clientConfig) error {
+		c.acl = acl
+		return nil
+	}
+}
+
+// WithConnectTimeout returns a configOption specifying a non-default connection
+// timeout when we try to establish a connection to a ZooKeeper server.
+func WithConnectTimeout(t time.Duration) configOption {
+	return func(c *clientConfig) error {
+		if t.Seconds() < 1 {
+			return errors.New("Invalid Connect Timeout. Minimum value is 1 second")
```

Comment:
"[Error strings should not be capitalized (unless beginning with proper nouns or acronyms) or end with punctuation, since they are usually printed following other context](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)."